### PR TITLE
Allow specifying a NamePrefix for temp files (BL-10005)

### PR DESCRIPTION
**Pull request by JohnThomson**
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156)

### CODE REVIEW

**Review by andrew-polk — CHANGES_REQUESTED** [review]
Needs semver:minor.
And an addition to changelog.md

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-:-N-4N0jY5IN7GDEix9cj:bsesjpx)* status: all files reviewed, 2 unresolved discussions (waiting on @JohnThomson)

___
*[SIL.Core/IO/TempFile.cs, line 60 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4OoPt6AQe0Ii1j_p1:-N-4OoPt6AQe0Ii1j_p2:b7jypbs) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core/IO/TempFile.cs#L60)):*
> ```Smalltalk
> 			while (needMoreTries)
> 			{
> 				result = System.IO.Path.GetTempPath() + (NamePrefix??"") + System.IO.Path.GetRandomFileName();
> ```

Is Path.Combine better here?
___
*[SIL.Core.Tests/IO/TempFileTests.cs, line 153 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4Pa0F9gKVg_rX2EH-:-N-4Pa0F9gKVg_rX2EH0:b-1wol3u) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core.Tests/IO/TempFileTests.cs#L153)):*
> ```Smalltalk
> 			Assert.That(RobustFile.Exists(temp1.Path), Is.False);
> 			Assert.That(Directory.Exists(temp2.Path), Is.False);
> 			TempFile.NamePrefix = null;
> ```

Do we need to do anything to ensure this fixture is single-threaded? (So that no other test tries to run while one of these has NamePrefix set?)

**Comment by tombogle** [comment]
*[SIL.Core/IO/TempFile.cs, line 120 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4QhcL-13gIS7udwSX:-N-4QhcL-13gIS7udwSY:bt1xedo) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core/IO/TempFile.cs#L120)):*
> ```Smalltalk
> 		/// Try to use a clearly unique prefix so you don't conflict with other apps.
> 		/// </summary>
> 		public static string NamePrefix { get; set; }
> ```

Should the comments for the other constructors include a "see" reference to this static property?

**Review by JohnThomson — COMMENTED** [review]
done

> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-:-N-4VOMQ0TZIGhK_AhUO:b-959hsp)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @andrew-polk and @tombogle)

___
*[SIL.Core/IO/TempFile.cs, line 60 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4OoPt6AQe0Ii1j_p1:-N-4Vho-9KqdLCn3zeiE:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core/IO/TempFile.cs#L60)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Is Path.Combine better here?
</blockquote></details>

Done.
___
*[SIL.Core/IO/TempFile.cs, line 120 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4QhcL-13gIS7udwSX:-N-4Wmel0wagTaNm902Y:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core/IO/TempFile.cs#L120)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

Should the comments for the other constructors include a "see" reference to this static property?
</blockquote></details>

Done.
___
*[SIL.Core.Tests/IO/TempFileTests.cs, line 153 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4Pa0F9gKVg_rX2EH-:-N-4Z-uj49uOVRk_7CPK:bqxhpjq) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core.Tests/IO/TempFileTests.cs#L153)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Do we need to do anything to ensure this fixture is single-threaded? (So that no other test tries to run while one of these has NamePrefix set?)
</blockquote></details>

I think what I've done will ensure that.

**Review by andrew-polk — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156)* status: all files reviewed, 2 unresolved discussions (waiting on @JohnThomson and @tombogle)

___
*[SIL.Core.Tests/IO/TempFileTests.cs, line 153 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4Pa0F9gKVg_rX2EH-:-N-4l_fY9gjWwYS6wbGA:baz01d3) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core.Tests/IO/TempFileTests.cs#L153)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think what I've done will ensure that.
</blockquote></details>

Thanks. One more (`WithNamePrefix_CreatesFileStartingWithThat()`) though I see now that Nunit claims it doesn't ever run parallel by default, so I guess these weren't needed.

**Review by JohnThomson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @andrew-polk and @tombogle)

___
*[SIL.Core.Tests/IO/TempFileTests.cs, line 153 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4Pa0F9gKVg_rX2EH-:-N-52lwS121YI07uAriR:bb7f2mf) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core.Tests/IO/TempFileTests.cs#L153)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Thanks. One more (`WithNamePrefix_CreatesFileStartingWithThat()`) though I see now that Nunit claims it doesn't ever run parallel by default, so I guess these weren't needed.
</blockquote></details>

done

**Comment by tombogle** [comment]
*[SIL.Core/IO/TempFile.cs, line 120 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4QhcL-13gIS7udwSX:-N-8X5rg9WFdScEIlh_u:bz69hcc) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core/IO/TempFile.cs#L120)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Although I don't feel strongly about it, I kind of prefer using the <see...> element in these comments because it makes it possible to Ctrl-click to go to the thing being referenced.

**Review by andrew-polk — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156)* status: all files reviewed, 1 unresolved discussion (waiting on @JohnThomson)

**Review by JohnThomson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @andrew-polk and @tombogle)

___
*[SIL.Core/IO/TempFile.cs, line 120 at r1](https://reviewable.io/reviews/sillsdev/libpalaso/1156#-N-4QhcL-13gIS7udwSX:-N-9WdN5EIRujiFnf6wV:b-o50r5q) ([raw file](https://github.com/sillsdev/libpalaso/blob/e74f57a0aff516c69dd885c054b6622a145f1225/SIL.Core/IO/TempFile.cs#L120)):*
<details><summary><i>Previously, tombogle (Tom Bogle) wrote…</i></summary><blockquote>

Although I don't feel strongly about it, I kind of prefer using the <see...> element in these comments because it makes it possible to Ctrl-click to go to the thing being referenced.
</blockquote></details>

I think I've done what you want...I'm not very familiar with this mechanism.

**Review by andrew-polk — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/libpalaso/1156)* status: all files reviewed, 1 unresolved discussion (waiting on @tombogle)

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 
 ### Added
-
+- [SIL.Core.IO] NamePrefix setting and CleanupTempFolder method added to TempFile
```
Comment (andrew-polk) [inline CHANGELOG.md:20]:
```suggestion
- [SIL.Core] NamePrefix setting and CleanupTempFolder method added to TempFile
```

**Pull request merged**